api/agent/drivers/docker/docker.go

```diff
@@ -102,13 +164,41 @@ func NewDocker(conf drivers.Config) *DockerDriver {
 		}
 	}
 
+	liopts := docker.ListImagesOptions{All: false}
+	requiredImages := ListOfImageIDs(conf.RequiredImages)
+	liopts.Context = context.Background()
+
 	if conf.DockerLoadFile != "" {
 		err = loadDockerImages(driver, conf.DockerLoadFile)
 		if err != nil {
 			logrus.WithError(err).Fatalf("cannot load docker images in %s", conf.DockerLoadFile)
 		}
 	}
 
+	if conf.MaxImageCacheSize != 0 {
+		driver.imageCache = NewCache()
+
+		go func(context context.Context) {
+			liopts := docker.ListImagesOptions{All: false}
+			liopts.Context = context
+			images, err := driver.docker.ListImages(liopts)
+
+			if err != nil {
+				logrus.WithError(err).Fatalf("cannot list docker images %s", err)
+			}
+			for _, i := range images {
+				if Contains(requiredImages, i) {
+					driver.imageCache.Add(i)
+				} else {
+					driver.imageCache.Add(i)
+					driver.imageCache.Lock(i.ID, "baseimage")
+				}
+			}
+		}(context.Background())
+
+		go NewImageCleaner(context.Background(), driver, int64(conf.MaxImageCacheSize))
```

Comment:
I think we should start this go-routine after the list processing above is completed, maybe at line 196?